python/python/lance/tf/data.py

```diff
@@ -0,0 +1,229 @@
+
+
+"""Tensorflow Dataset (`tf.data <https://www.tensorflow.org/guide/data>`_)
+implementation for Lance.
+
+.. warning::
+
+    Experimental feature. API stability is not guaranteed.
+"""
+
+from __future__ import annotations
+
+import logging
+from pathlib import Path
+from typing import Dict, Iterable, List, Optional, Union
+
+import lance
+import numpy as np
+import pyarrow as pa
+import tensorflow as tf
+from lance import LanceDataset
+from lance.fragment import LanceFragment
+
+
+def arrow_data_type_to_tf(dt: pa.DataType) -> tf.DType:
+    """Convert Pyarrow DataType to Tensorflow."""
+    if pa.types.is_boolean(dt):
+        return tf.bool
+    elif pa.types.is_int8(dt):
+        return tf.int8
+    elif pa.types.is_int16(dt):
+        return tf.int16
+    elif pa.types.is_int32(dt):
+        return tf.int32
+    elif pa.types.is_int64(dt):
+        return tf.int64
+    elif pa.types.is_uint8(dt):
+        return tf.uint8
+    elif pa.types.is_uint16(dt):
+        return tf.uint16
+    elif pa.types.is_uint32(dt):
+        return tf.uint32
+    elif pa.types.is_uint64(dt):
+        return tf.uint64
+    elif pa.types.is_float16(dt):
+        return tf.float16
+    elif pa.types.is_float32(dt):
+        return tf.float32
+    elif pa.types.is_float64(dt):
+        return tf.float64
+    elif (
+        pa.types.is_string(dt)
+        or pa.types.is_large_string(dt)
+        or pa.types.is_binary(dt)
+        or pa.types.is_large_binary(dt)
+    ):
+        return tf.string
+
+    raise TypeError(f"Arrow/Tf conversion: Unsupported arrow data type: {dt}")
+
+
+def data_type_to_tensor_spec(dt: pa.DataType) -> tf.TensorSpec:
+    """Convert PyArrow DataType to Tensorflow TensorSpec."""
+    if (
+        pa.types.is_boolean(dt)
+        or pa.types.is_integer(dt)
+        or pa.types.is_floating(dt)
+        or pa.types.is_string(dt)
+    ):
+        return tf.TensorSpec(shape=(None,), dtype=arrow_data_type_to_tf(dt))
+    elif pa.types.is_fixed_size_list(dt):
+        return tf.TensorSpec(
+            shape=(None, dt.list_size), dtype=arrow_data_type_to_tf(dt.value_type)
+        )
+    elif pa.types.is_list(dt) or pa.types.is_large_list(dt):
+        return tf.TensorSpec(
+            shape=(
+                None,
+                None,
+            ),
+            dtype=arrow_data_type_to_tf(dt.value_type),
+        )
+
+    raise TypeError("Unsupported data type: ", dt)
+
+
+def schema_to_spec(schema: pa.Schema) -> tf.TypeSpec:
+    """Convert PyArrow Schema to Tensorflow output signature."""
+    signature = {}
+    for name in schema.names:
+        field = schema.field(name)
+        signature[name] = data_type_to_tensor_spec(field.type)
+    return signature
+
+
+def column_to_tensor(column: pa.Array, tensor_spec: tf.TensorSpec) -> tf.Tensor:
+    if isinstance(tensor_spec, tf.RaggedTensorSpec):
+        return tf.ragged.constant(column, dtype=tensor_spec.dtype)
+    else:
+        return tf.constant(column, dtype=tensor_spec.dtype)
```

Comment:
Are there any docs on how this works? We can just pass arrow arrays into `tf.constant` and tensorflow understands arrow? Or is there some other protocol here?

Reply:
https://www.tensorflow.org/tutorials/customization/basics#numpy_compatibility it is not yet clear to me that this is a zero copy operation yet.

Let's digg more in implementation details as follow up? Wanted to get the API spec out for now.